nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp

```diff
@@ -275,6 +275,22 @@ class BtActionNode : public BT::ActionNodeBase
       }
     }
 
+    auto start_time = node_->now();
+    while (rclcpp::ok()
+           && !goal_result_available_
+           && node_->now() - start_time < server_timeout_) {
```

Comment:
This part of the condition `node_->now() - start_time < server_timeout_` basically is just giving the cancel future only 1 more set of `server_timeout_` to get a cancel. The spin until future complete in L269 will block for `server_timeout_` waiting on this while loop will also block for another set of `server_timeout_` but is that really what you want to do? Its just giving that now 2 chances of that same length to do so, so you could still run into this issue if it exceeds that time frame. 

I don't think you actually really care about the `goal_result_available_` as much as you care that the `cancel_future` is completed, indicating that the server actually canceled successfully.

Reply:
What is your `server_timeout` set to? We did increase this to I believe 20ms up from 10ms a few months ago when this logic for the server timeout was changed to enable the BT loop duration to move at a consistent rate

Reply:
You mention 

> This will force the result_callback to be executed after the sub-action is cancelled, allowing to get a clean BtNavigator's action result.

and point to the `spin_some()`. So is some of this that you actually just want to spin to get a result callback more than anything else? If that' true, we could add a `spin_some()` to `halt()` after cancellation as well. However if the return code on the `future_cancel` is `SUCCESS` then we know it triggered so just 1 `spin_some()` should do the trick without having to do any of the rest of it.

Reply:
> This part of the condition `node_->now() - start_time < server_timeout_` basically is just giving the cancel future only 1 more set of `server_timeout_` 

Yes, that's right ...

> to get a cancel

... but no, not a cancel. This condition does not aim at getting the **cancel** result, but the **goal** result.
In my understanding, here is what happen when entering the `halt()` function:
1) First step of the `halt()` function, already implemented: The **cancel** request is sent to the sub-action. For a time period of maximum `server_timeout_`, we wait for the **cancel** result. Considering we have no trouble for getting the **cancel** future, there are two possibilities at the end of this first *set of `server_timeout_`*:
  - Case A: The **cancel** result and the **goal** result are available
  - Case B: Only the **cancel** result is result is available. Actually, in my opinion, this is not a exceptional case at all, since the sub-action can automatically accept the cancel requests (so the **cancel** result will be "instantaneously" available after sending the **cancel** request), but this same sub-action can take some time to properly end/stop the on-going **goal** after the cancel acceptation (so the **goal** result will not be available right after the **cancel** acceptation). I think it depends on the behavior you want your sub-action to have in case of a (successful) cancel.
  
2) Second step of the `halt()` function, object of this PR: We want to check for the **goal** result availability before ending the halt of the BT. This second step depends on the case obtained at step 1:
- If Case A: we don't even enter the while loop since the **goal** result is already available.
- If Case B: we enter the while loop. For a time period of maximum `server_timeout_`, we wait for the **goal** result.

So I think that waiting for maximum two *set of `server_timeout_`* for properly ending the sub-action in case of successful cancel is:
- neither only useful for isolated use cases,
- neither too much cautious / a possible slow down cause,

since it strongly depends on the desired behavior of the sub-action when it is being cancelled, and since it seems normal to me to wait for the sub-action to properly ends before halting the BT.

Reply:
> What is your `server_timeout` set to? We did increase this to I believe 20ms up from 10ms a few months ago when this logic for the server timeout was changed to enable the BT loop duration to move at a consistent rate

In our use case, we set the `server_timeout` to 100 ms.

Reply:
> So is some of this that you actually just want to spin to get a result callback more than anything else?

From a behavior point of view, what we *want more than anything else* is the sub-action to be cleanly finished before halting the BT. To know if it is indeed cleanly finished, we check on the availability of the sub-action **goal** result, since the **goal** result is available only once the sub-action has ended. Once the sub-action **goal** result is available, the goal result callback is automatically executed. So yes, saying that what we *want more than anything else* is the sub-action to be cleanly finished, is actually saying we *want to spin to get a result callback more than anything else*.

> However if the return code on the future_cancel is `SUCCESS` then we know it triggered so just 1 `spin_some()` should do the trick without having to do any of the rest of it.

No I don't agree with this, there is an explanation in this [previous comment](https://github.com/ros-planning/navigation2/pull/2436#discussion_r665448357)

> As far as I can tell from this PR there are 2 distinct issues you're trying to address (or just 1 and I'm misinterpreting for the second):
> 
>     * `cancel_future` doesn't complete in `server_timeout_` so you want to spin for a bit longer until it actually does successfully (or is that not an issue you're running into, with that error log?)
> 
>     * Because you don't spin after canceling, the callback for result isn't triggered in that same BT run so its messing up the state for the next run, so you're adding in a `spin_some()` to process the cancel result callback before exiting `halt()` to clean things up for the next run.
> If only (B) is your issue, no changes are required to the code above this block and this block can be largely simplified without the `while` to just be `spin_some()` since if `cancel_future` is successful, you're promised to have that waiting in your callback queue. If (A) is your issue, then we should work on combining what you have now with the while loop with spinning the future.

Well, I would said only (B) is an issue, but not exactly. I would reformulate it this way to make it accurate (changes are in bold):
Because you don't spin after canceling, the callback for **goal** result isn't triggered in that same BT run so its messing up the state for the next run, so you're adding in a `spin_some()` to process the **goal (Not cancel)** result callback before exiting `halt()` to clean things up for the next run.

So, I don't really agree with your conclusion about only (B) being an issue. If the `cancel_future` is successful, it doesn't mean the sub-action will have enough time to finish cleanly before halting the BT. It only means that the **cancel** request has been accepted. Which is great, but not enough, since the sub-action might not stop the current **goal** right after having accepting the **cancel** request.

Reply:
Got it -- see comment in the thread above. I think we're on the same page, but I think we can simplify this if we leverage some internal knowledge about how the actions work. If that turns out not to be true, then we can discuss what to do so we can make sure we process the goal callback